lib/parser.php

```diff
@@ -266,14 +266,37 @@ private function peg_f3($blockName, $attrs) {
         );
         }
     private function peg_f4($s, $children, $e) {
-        list( $innerHTML, $innerBlocks ) = peg_array_partition( $children, 'is_string' );
-
-        return array(
+	    $innerHTMLBeforeInnerBlocks = array();
+	    $innerHTMLAfterInnerBlocks = array();
+	    $innerBlocks = array();
+
+	    $inner_found = false;
+	    foreach ( $children as $child ) {
+	    	if( is_string( $child ) ) {
+	    		if ( $inner_found ) {
+	    			$innerHTMLAfterInnerBlocks[] = $child;
+			    } else {
+				    $innerHTMLBeforeInnerBlocks[] = $child;
+			    }
+		    } else {
+	    		$innerBlocks[] = $child;
+	    		$inner_found = true;
+		    }
+	    }
+
+
+        $block = array(
           'blockName'  => $s['blockName'],
           'attrs'      => $s['attrs'],
           'innerBlocks'  => $innerBlocks,
-          'innerHTML'  => implode( '', $innerHTML ),
+          'innerHTML'  => implode( '', $innerHTMLBeforeInnerBlocks ) . implode( '', $innerHTMLAfterInnerBlocks ),
         );
+
+	    if( $inner_found ) {
+          $block['innerHTMLBeforeInnerBlocks']  = implode( '', $innerHTMLBeforeInnerBlocks );
+          $block['innerHTMLAfterInnerBlocks']  = implode( '', $innerHTMLAfterInnerBlocks );
+	    }
+	    return $block;
```

Comment:
hm…it seems like we are trying to manually edit automatically-generated code here. @brucepearson did you get a chance to look at the spec grammar which is run through `php-pegjs` to generate this file?

Reply:
Thanks @dmsnell. Is the concept OK that I've applied here? I can then translate to the grammar file.